text/0000-dynamic-import-hints.md

```diff
@@ -0,0 +1,264 @@
+            specifier: {
+                id: '@salesforce/client/formFactor',
+                namespacedId: undefined, // scoped modules are namespaced at compile time
+                type: 'client',
+                file: 'foo.js',
+                location: {...}
+            },
+            value: 'SMALL',
+        },
+        {
+            rawValue: '/* "@salesforce/userPermission/ViewSetup": "true" */'
+            specifier: {
+                id: '@salesforce/userPermission/ViewSetup',
+                namespacedId: undefined, // scoped modules are namespaced at compile time
+                type: 'userPermission',
+                file: 'foo.js',
+                location: {...}
+            },
+            value: 'true',
+        }]
+    }]
+}
+```
+
+## Other Options Considered
+
+### Hints as a query attached to the component name
+The hint syntax is a parameterized query appended to the component name:
+```js
+export async function example1() {
+    loadedModule = await import("lightning-foo*?*@salesforce/client/formFactor=Small");
+}
+```
+
+*Pros:*
+
+* Intuitive syntax
+
+*Cons:*
+
+* Requires transform to remove the query
+* Using a query string syntax to discover modules to prefetch doesn't align with the way standard import() works. JavaScript VM caches modules by URL (including the query string), this means that lightning-foo lightning-foo?1 and lightning-foo?2 are 3 different JavaScript module instances
+
+### Hints via proprietary LWC function
+The hint syntax is expressed via a proprietary lwc function, where the first parameter is a component name and the second parameter is a query object - hintImport(cmpName, queryObj). At compile time, the function will be transformed into a regular import without a string. Note, the hint is consumed by metadata parser only, which is traversed outside of the compilation sequence on the original source. Therefore, it is safe to have a ‘throw-away’ wrapper that will be eliminated at compiler time.  
+
+```js
+import formFactor from '@salesforce/client/formFactor';
+import permName from '@salesforce/customPermission/PermName';
+
+export async function example1() {
+    loadedModule = await hintImport(
+        "lightning-foo",
+        {
+            query: "{0} IS Small AND {1} NOT true",
+            fields: [{
+                resource: formFactor,
+                value: "Small",
+            }, {
+                resource: permName,
+                value: true
+            }]
+        }
+    );
+}
+```
+
+*Pros:*
+
+* Dedicated syntax
+* Full control of the hint shape
+* Having to explicitly define a resource reference will allow for ref integrity support if required
+
+*Cons:*
+
+* Requires an additional transform
+* Requires knowledge of syntax existence
+* Maintanance burden
+
+## Future work
+* Supporting hints for static imports(`import Foo from "lightning-foo"; /* "@salesforce/client/formFactor": "SMALL" */`)
```

Comment:
Why, static imports are runtime independent.

Reply:
It could be as a way to override a backend configuration and force to load a specific variation of a module.
For example: Consider there is an application framework which uses AMD format and injects dependencies. Based on a runtime condition, it could be injecting a mobile version of the dependencies. If the client chooses to override that for a specific case, it can be done with a hint to force the desktop version of the dependency. 
Something along those lines. There is no such usecase yet, but was leaving that option open.

I could update the language to say `Support for hints in static imports is not in consideration at this point.`

Reply:
Something like [import maps](https://github.com/salesforce/lws-rfcs/pull/16) may be a better tool for that, particularly since the hint value doesn't vary across the lifetime of the `document`.